pkg/tbtc/node.go

```diff
@@ -205,3 +268,136 @@ func (n *node) joinDKGIfEligible(seed *big.Int, startBlockNumber uint64) {
+	for _, misbehavedID := range dkgResultEvent.Misbehaved {
+		misbehavedSet[misbehavedID] = struct{}{}
+	}
+
+	// If member is considered as misbehaved, it could not stay in the group.
+	if _, isMisbehaved := misbehavedSet[memberIndex]; isMisbehaved {
+		return nil, fmt.Errorf(
+			"[member:%v] could not stay in the group because "+
+				"the member is considered as misbehaving",
+			memberIndex,
+		)
+	}
+
+	// Construct a new view of the operating members according to the accepted
+	// DKG result.
+	operatingMemberIDs := make([]group.MemberIndex, 0)
+	for _, memberID := range result.Group.MemberIDs() {
+		if _, isMisbehaved := misbehavedSet[memberID]; !isMisbehaved {
+			operatingMemberIDs = append(operatingMemberIDs, memberID)
+		}
+	}
+
+	return operatingMemberIDs, nil
+}
+
+// waitForDkgResultEvent waits for the DKG result submission event. It times out
+// and returns error if the DKG result event is not emitted on time.
+func (n *node) waitForDkgResultEvent(
+	dkgResultChannel chan *DKGResultSubmittedEvent,
+	publicationStartBlock uint64,
+) (*DKGResultSubmittedEvent, error) {
+	config := n.chain.GetConfig()
+
+	timeoutBlock := publicationStartBlock + dkg.PrePublicationBlocks() +
+		(uint64(config.GroupSize) * config.ResultPublicationBlockStep)
+
+	blockCounter, err := n.chain.BlockCounter()
+	if err != nil {
+		return nil, err
+	}
+
+	timeoutBlockChannel, err := blockCounter.BlockHeightWaiter(timeoutBlock)
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	case dkgResultEvent := <-dkgResultChannel:
+		return dkgResultEvent, nil
+	case <-timeoutBlockChannel:
+		return nil, fmt.Errorf("ECDSA DKG result publication timed out")
+	}
+}
+
+// resolveSigningGroupOperators takes two parameters:
+// - selectedOperators: Contains addresses of all selected operators. Slice
+//   length equals to the groupSize. Each element with index N corresponds
+//   to the group member with ID N+1.
+// - operatingGroupMembersIDs: Contains group members IDs that were neither
+//   disqualified nor marked as inactive. Slice length is lesser than or equal
+//   to the groupSize.
+//
+// Using those parameters, this function transforms the selectedOperators
+// slice into another slice that contains addresses of all operators
+// that were neither disqualified nor marked as inactive. This way, the
+// resulting slice has only addresses of properly operating operators
+// who form the resulting group.
+//
+// Example:
+// selectedOperators: [member1, member2, member3, member4, member5]
+// operatingGroupMembersIDs: [5, 1, 3]
+// signingGroupOperators: [member1, member3, member5]
+func (n *node) resolveSigningGroupOperators(
+	selectedOperators []chain.Address,
+	operatingGroupMembersIDs []group.MemberIndex,
+) ([]chain.Address, error) {
+	config := n.chain.GetConfig()
+
+	if len(selectedOperators) != config.GroupSize ||
+		len(operatingGroupMembersIDs) < config.HonestThreshold {
+		return nil, fmt.Errorf("invalid input parameters")
+	}
```

Comment:
In practice, we will have a `GroupQuorum` parameter (see #3169) so we should check against it instead of the `HonestThreshold`. Let's leave a TODO about changing it.

Reply:
Done in 54916c43296c6911d2229b5ad1af34509c3d77cf.

Reply:
Nice! Leaving it unresolved for reference. I made a note about addressing it in #3041